### PR TITLE
Don't delete the pointer we reference to

### PR DESCRIPTION
**Pull request by adeas31**

### CODE REVIEW

File: src/OMSimulatorLib/oms2/Element.cpp [hunk 1]
```diff
@@ -90,10 +90,11 @@ oms2::Element& oms2::Element::operator=(const oms2::Element& rhs)
   for (int i=0; i<nInterfaces; ++i)
     this->interfaces[i] = reinterpret_cast<oms_connector_t*>(new oms2::Connector(*reinterpret_cast<oms2::Connector*>(rhs.interfaces[i])));
 
+  if (this->geometry)
+    delete reinterpret_cast<oms2::ssd::ElementGeometry*>(this->geometry);
   oms2::ssd::ElementGeometry* geometry_ = new oms2::ssd::ElementGeometry();
   *geometry_ = *reinterpret_cast<oms2::ssd::ElementGeometry*>(rhs.geometry);
   this->geometry = reinterpret_cast<ssd_element_geometry_t*>(geometry_);
-  delete geometry_;
```
Comment (lochel) [inline src/OMSimulatorLib/oms2/Element.cpp]: You are right, this need to be fixed. Maybe it is better to do it like this:
```c++
if (this->geometry)
  *reinterpret_cast<oms2::ssd::ElementGeometry*>(this->geometry) = *reinterpret_cast<oms2::ssd::ElementGeometry*>(rhs.geometry);
else
  this->geometry = reinterpret_cast<ssd_element_geometry_t*>(new oms2::ssd::ElementGeometry(reinterpret_cast<oms2::ssd::ElementGeometry>(*rhs.geometry)));
```

File: src/OMSimulatorLib/oms2/Element.cpp [hunk 2]
```diff
@@ -117,7 +118,6 @@ void oms2::Element::setGeometry(const oms2::ssd::ElementGeometry* newGeometry)
   {
     oms2::ssd::ElementGeometry* geometry_ = new oms2::ssd::ElementGeometry(*newGeometry);
     this->geometry = reinterpret_cast<ssd_element_geometry_t*>(geometry_);
-    delete geometry_;
```
Comment (lochel) [inline src/OMSimulatorLib/oms2/Element.cpp]: And here the same:

```c++
if (this->geometry)
  *reinterpret_cast<oms2::ssd::ElementGeometry*>(this->geometry) = *reinterpret_cast<oms2::ssd::ElementGeometry*>(newGeometry);
else
  this->geometry = reinterpret_cast<ssd_element_geometry_t*>(new oms2::ssd::ElementGeometry(*newGeometry));
```

File: src/OMSimulatorLib/oms2/Element.cpp (hunk 1 above)
Comment (adeas31) [reply src/OMSimulatorLib/oms2/Element.cpp]: I think I tested it like that but for some reason the following doesn't call the assignment operator,
`  *reinterpret_cast<oms2::ssd::ElementGeometry*>(this->geometry) = *reinterpret_cast<oms2::ssd::ElementGeometry*>(rhs.geometry);`

Comment (lochel) [reply src/OMSimulatorLib/oms2/Element.cpp]: Weird. I will test it tomorrow.

Comment (lochel) [reply src/OMSimulatorLib/oms2/Element.cpp]: Do you have an example that is calling `oms2::Element::operator=`?

Comment (adeas31) [reply src/OMSimulatorLib/oms2/Element.cpp]: No. I expect the above line of code to call it.

Comment (lochel) [reply src/OMSimulatorLib/oms2/Element.cpp]: No, that one should call `oms2::ssd::ElementGeometry::operator=`.

Comment (adeas31) [reply src/OMSimulatorLib/oms2/Element.cpp]: Ahhh of course. I didn't looked at it closely.
So why do you need an example for `oms2::Element::operator=`?

Comment (lochel) [reply src/OMSimulatorLib/oms2/Element.cpp]: I want to test test the changes.

Comment (adeas31) [reply src/OMSimulatorLib/oms2/Element.cpp]: I think call `oms2_getElements` and then `oms2_unloadModel` and it crash.

Comment (lochel) [reply src/OMSimulatorLib/oms2/Element.cpp]: `oms2::Element::operator=` is not used at all and actually shouldn’t be even implemented.

The crash is caused by `oms2::Element::setGeometry` which can be simplified to:
```c++
void oms2::Element::setGeometry(const oms2::ssd::ElementGeometry* newGeometry)
{
  if (this->geometry)
  {
    delete reinterpret_cast<oms2::ssd::ElementGeometry*>(this->geometry);
    this->geometry = NULL;
  }

  if (newGeometry)
  {
    this->geometry = reinterpret_cast<ssd_element_geometry_t*>(new oms2::ssd::ElementGeometry(*newGeometry));
  }
}
```

Comment (lochel) [reply src/OMSimulatorLib/oms2/Element.cpp]: I added a test case for this, which can be activated as soon as this is fixed: https://github.com/OpenModelica/OMSimulator-testsuite/commit/f90bcd5f798499f7ecd232e70b5cd0aa58503472

**Pull request merged**